File: ports/go/src/core/io/bcstextstream.go
package io

import (
	"bcs/core/global"
	"bufio"
	"io"
	"os"
)

// BcsTextStream provides a convenient interface for reading and writing text.
type BcsTextStream struct {
	reader *bufio.Reader
	writer *bufio.Writer
}

func NewBcsTextStream(file *BcsFile) *BcsTextStream {
	// Assume BcsFile has an underlying os.File accessible or passed
	// This is a conceptual map for the port
	return &BcsTextStream{
		reader: bufio.NewReader(file.file),
		writer: bufio.NewWriter(file.file),
	}
}

func (s *BcsTextStream) ReadLine() (global.BcsString, error) {
	line, err := s.reader.ReadString('\n')
	if err != nil && err != io.EOF {
		return "", err
	}
	return global.NewBcsString(line), nil
}

func (s *BcsTextStream) ReadAll() (global.BcsString, error) {
	bytes, err := io.ReadAll(s.reader)
	if err != nil {
		return "", err
	}
	return global.NewBcsString(string(bytes)), nil
}

func (s *BcsTextStream) WriteString(str global.BcsString) error {
	_, err := s.writer.WriteString(string(str))
	if err != nil {
		return err
	}
	return s.writer.Flush()
}

File: ports/rust/src/core/io/bcstextstream.rs
use bcs_core::global::bcsstring::BcsStringWrapper;
use std::fs::File;
use std::io::{self, BufRead, BufReader, Write};

// BcsTextStream provides a convenient interface for reading and writing text.
pub struct BcsTextStream {
    reader: Option<BufReader<File>>,
    writer: Option<File>,
}

impl BcsTextStream {
    pub fn new(file: File) -> Self {
        // Conceptually maps the stream to underlying file handles
        let file_clone = file.try_clone().unwrap(); // Simplification for port mapping
        BcsTextStream {
            reader: Some(BufReader::new(file)),
            writer: Some(file_clone),
        }
    }

    pub fn read_line(&mut self) -> io::Result<BcsStringWrapper> {
        let mut buffer = String::new();
        if let Some(ref mut reader) = self.reader {
            reader.read_line(&mut buffer)?;
        }
        Ok(BcsStringWrapper::new(&buffer))
    }

    pub fn write_string(&mut self, str: &BcsStringWrapper) -> io::Result<()> {
        if let Some(ref mut writer) = self.writer {
            writer.write_all(str.to_utf8().as_slice())?;
            writer.flush()?;
        }
        Ok(())
    }
}
